### PR TITLE
`clark-service-refactor` to `main`

### PR DESCRIPTION
**Pull request by mfranz1**
This PR updates `clark-client` to conform to changes in clark-services. All services have been rewritten except for `learning-object-service`. All LOS routes are still currently being used to the old service and are referred to as legacy routes in the client.

Majority of file changes are files that have been reorganized into subdirectories of `src/app/core`. The core module now reflects our clark-service/CLARK-Gateway strutcures.

This pull request is linked to [Shortcut Story #28112: Update clark-client to reflect changes in clark-service](https://app.shortcut.com/clarkcan/story/28112/update-clark-client-to-reflect-changes-in-clark-service).

### CODE REVIEW

File: src/app/core/feature-module/featured.service.ts [hunk 1]
```diff
@@ -130,22 +133,22 @@ export class FeaturedObjectsService {
    * @returns {Promise<LearningObject[]>}
    * @memberof LearningObjectService
    */
-  getNotFeaturedLearningObjects(query?: Query): Promise<{learningObjects: LearningObject[], total: number}> {
+  getNotFeaturedLearningObjects(query?: Query): Promise<{ learningObjects: LearningObject[], total: number }> {
```
[the remark this answers is not in the document]
Comment (Cwagne17) [reply src/app/core/feature-module/featured.service.ts:136]: Is this actually used somewhere.

File: src/app/core/feature-module/featured.service.ts [hunk 2]
```diff
@@ -165,9 +168,8 @@ export class FeaturedObjectsService {
    */
   getCollectionFeatured(collection: string) {
     return this.http
-      .get(FEATURED_ROUTES.GET_COLLECTION_FEATURED(collection))
+      .get(LEGACY_FEATURED_ROUTES.GET_COLLECTION_FEATURED(collection))
```
Comment (Cwagne17) [inline src/app/core/feature-module/featured.service.ts:171]: This should use get featured objects with the query collection.

File: src/app/core/library-module/library.routes.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    return `${environment.apiURL}/users/${encodeURIComponent(username)}/library`;
+  },
+  /**
+   * Request to remove a learning object from a user's library
+   * @method DELETE
+   * @auth required
+   * @param username username of the user's library
+   * @param learningObjectId cuid of the learning object to remove
+   */
+  REMOVE_LEARNING_OBJECT_FROM_LIBRARY(username, learningObjectId) {
```
Comment (Cwagne17) [inline src/app/core/library-module/library.routes.ts:30]: The backedn is expecting the libraryId not learningObjectId

File: src/app/core/library-module/library.service.ts [hunk 4]
```diff
@@ -86,14 +89,14 @@ export class LibraryService {
     }
     this.http
       .delete(
-        USER_ROUTES.CLEAR_LEARNING_OBJECT_FROM_CART(
+        LIBRARY_ROUTES.REMOVE_LEARNING_OBJECT_FROM_LIBRARY(
           this.user.username,
           cuid
```
Comment (Cwagne17) [inline src/app/core/library-module/library.service.ts:94]: The backend is expecting the libraryId not cuid. Also in the routes file it says learningObjectId

File: src/app/core/library-module/library.service.ts [hunk 5]
```diff
@@ -67,7 +70,7 @@ export class LibraryService {
     }
     return await this.http
       .post(
-        USER_ROUTES.ADD_LEARNING_OBJECT_TO_CART(
+        LIBRARY_ROUTES.ADD_LEARNING_OBJECT_TO_LIBRARY(
           this.user.username
         ),
         {
```
Comment (Cwagne17) [inline src/app/core/library-module/library.service.ts:76]: In the request body authorUsername is not used on the backend.

File: src/app/core/library-module/library.service.ts [hunk 6]
```diff
@@ -113,7 +116,7 @@ export class LibraryService {
     // Show loading spinner
     this._loading$.next(true);
     // Url route for bundling
-    const bundle = USER_ROUTES.OBJECT_BUNDLE(
+    const bundle = BUNDLING_ROUTES.BUNDLE_LEARNING_OBJECT(
```
Comment (Cwagne17) [inline src/app/core/library-module/library.service.ts:119]: Why is bundling in library?

File: src/app/core/metric-module/metric.routes.ts [hunk 7]
```diff
@@ -0,0 +1,32 @@
+import { environment } from '@env/environment';
+
+export const METRIC_ROUTES = {
+  /**
+   * Request to get collection metrics
+   * @returns CollectionMetrics object containing metrics for the given collection
+   */
+  GET_COLLECTION_METRICS(abvCollectionName: string) {
+    return `${environment.apiURL}/metrics&collection=${abvCollectionName}`;
```
Comment (Cwagne17) [inline src/app/core/metric-module/metric.routes.ts:9]:
```suggestion
    return `${environment.apiURL}/metrics?collection=${abvCollectionName}`;
```

File: src/app/core/metric-module/metric.routes.ts [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+   * @returns UserMetrics object containing the total number of users and organizations
+   */
+  GET_USER_METRICS() {
+    return `${environment.apiURL}/users/metrics`;
+  },
+  /**
+   * Request to get all learning object stats
+   * @method GET
+   */
+  GET_LEARNING_OBJECT_STATS() {
```
Comment (Cwagne17) [inline src/app/core/metric-module/metric.routes.ts:29]: This isn't currently defined in the metrics module. This is probably legacy.

File: src/app/core/metric-module/metric.service.ts [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+import { throwError } from 'rxjs';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class MetricService {
+
+  constructor(private http: HttpClient) { }
+
+  getCollectionMetricsData(abvCollectionName: string) {
```
Comment (Cwagne17) [inline src/app/core/metric-module/metric.service.ts:14]: Where are the other four routes used?

File: src/app/core/rating-module/rating.service.ts [hunk 10]
```diff
@@ -251,31 +210,23 @@ export class RatingService {
...
    * @param {string} ratingId
    * @param {{concern: string, comment: string}} report
    * @returns {Promise<any>}
    */
   flagLearningObjectRating(params: {
-    username: string;
-    CUID: string;
-    version: number;
     ratingId: string;
     report: { concern: string, comment?: string }
```
Comment (Cwagne17) [inline src/app/core/rating-module/rating.service.ts:219]: Comment isn't used in the backend.

```suggestion
    report: { concern: string }
```

File: src/app/core/user-module/profiles.service.ts [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+  userNotifications: any;
+  constructor(private http: HttpClient, private auth: AuthService) { }
+
+  /**
+   * Function to retrieve a learning object
+   *
+   * @param params author is undefined - cuid is the current object cuid
+   * @returns a learning object with specified cuid
+   */
+  fetchLearningObject(params: { author: string; cuid: string }): Promise<any> {
```
Comment (Cwagne17) [inline src/app/core/user-module/profiles.service.ts:22]: The amount of getLearningObject service functions is crazy

File: src/app/core/user-module/user.routes.ts [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+    return `${environment.apiURL}/users`;
+  },
+  /**
+   * Request to retrieve a user's profile
+   * @method GET
+   * @auth required
+   * @param username the username of the user
+   * @returns the user's profile
+   */
+  GET_USER_PROFILE(username: string) {
```
Comment (Cwagne17) [inline src/app/core/user-module/user.routes.ts:36]: This route doesn't exist

File: src/app/core/user-module/user.service.ts [hunk 13]
```diff
@@ -56,17 +31,14 @@ export class UserService {
   ): Promise<void> {
     await this.http
       .put(
-        USER_ROUTES.ASSIGN_COLLECTION_MEMBER(collection, memberId),
+        LEGACY_USER_ROUTES.ASSIGN_COLLECTION_MEMBER(collection, memberId),
```
Comment (Cwagne17) [inline src/app/core/user-module/user.service.ts:34]: This should be access group module and I think we have a route that does this. We shouldn't be using legacy for permissions.

File: src/app/core/user-module/user.service.ts [hunk 14]
```diff
@@ -79,16 +51,17 @@ export class UserService {
...
    */
-  editMember(collection: string, memberId: string, role: string): Promise<User> {
+  editMember(
+    collection: string,
+    memberId: string,
+    role: string
+  ): Promise<User> {
     return this.http
-      .patch(USER_ROUTES.UPDATE_COLLECTION_MEMBER(collection, memberId), role, {
+      .patch(LEGACY_USER_ROUTES.UPDATE_COLLECTION_MEMBER(collection, memberId), role, {
```
Comment (Cwagne17) [inline src/app/core/user-module/user.service.ts:60]: Permissions should be in access group module and use the current routes that are implemented

File: src/app/core/user-module/user.service.ts [hunk 15]
```diff
@@ -107,39 +80,33 @@ export class UserService {
     await this.http
       .request(
         'delete',
-        USER_ROUTES.REMOVE_COLLECTION_MEMBER(collection, memberId),
+        LEGACY_USER_ROUTES.REMOVE_COLLECTION_MEMBER(collection, memberId),
```
Comment (Cwagne17) [inline src/app/core/user-module/user.service.ts:83]: Same here

File: src/app/core/user-module/user.service.ts [hunk 16]
```diff
@@ -107,39 +80,33 @@ export class UserService {
...
    * @param {string} query the text string to query by
-   * @returns {Promise<User[]} array of users matching the text query
+   * @returns {Promise<User[]>} array of users matching the text query
    * @memberof UserService
    */
   searchUsers(query: UserQuery): Promise<User[]> {
     return this.http
-      .get(USER_ROUTES.SEARCH_USERS(query), {
-        withCredentials: true
+      .get(LEGACY_USER_ROUTES.SEARCH_USERS(query), {
```
Comment (Cwagne17) [inline src/app/core/user-module/user.service.ts:102]: GET USERS should be used for search?

File: src/app/core/user-module/user.service.ts [hunk 17]
```diff
@@ -151,17 +118,14 @@ export class UserService {
    * @memberof UserService
    */
   getOrganizationMembers(organization: string): Promise<User[]> {
-    const route = USER_ROUTES.GET_SAME_ORGANIZATION(organization);
+    const route = LEGACY_USER_ROUTES.GET_SAME_ORGANIZATION(organization);
```
Comment (Cwagne17) [inline src/app/core/user-module/user.service.ts:121]: Is this used? Should we be adding this to clark-service?

File: src/app/core/user-module/user.service.ts [hunk 18]
```diff
@@ -186,85 +150,92 @@ export class UserService {
...
-    .toPromise();
-    this.getNotificationCount(username);
-    return deleteValue;
+  /**
+   * Grabs a users profile
+   *
+   * @param {string} username the username of the user to validate
+   * @returns {Promise<Object>} returns user name, email, and bio for profile meta data
+   */
+  fetchUserProfile(username: string): Promise<any> {
```
Comment (Cwagne17) [inline src/app/core/user-module/user.service.ts:207]: We don't have this route in clark-service

File: src/app/core/utility-module/utility.routes.ts [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+   */
+  GET_DOWNTIME() {
+    return `${environment.apiURL}/downtime?service=clark`;
+  },
+  /**
+   * Request to search organizations on CARD
+   * @method GET
+   * @returns list of organizations
+   */
+  SEARCH_ORGANIZATIONS(queryString: string) {
```
Comment (Cwagne17) [inline src/app/core/utility-module/utility.routes.ts:42]: Why is this here?

File: src/app/core/utility-module/utility.service.ts [hunk 20]
```diff
@@ -0,0 +1,73 @@
...
+  /**
+   * Gets all blogs from the database
+   *
+   * @returns An observable containing an array of blogs
+   */
+  getAllBlogs(): Observable<Blog[]> {
+    return this.http.get<Blog[]>(UTILITY_ROUTES.GET_BLOGS());
+  }
+
+  // /**
```
Comment (Cwagne17) [inline src/app/core/utility-module/utility.service.ts:24]: Is the create blog used?

File: src/app/core/utility-module/utility.service.ts [hunk 21]
```diff
@@ -0,0 +1,73 @@
...
+  //  * Gets all blogs from the database
+  //  *
+  //  * @returns An observable containing an array of blogs
+  //  */
+  // async getRecentBlogs(): Observable<Blog[]> {
```
Comment (Cwagne17) [inline src/app/core/utility-module/utility.service.ts:29]: How come this is commented out?

Clark-service supports the query recent

**Review by Cwagne17 — CHANGES_REQUESTED** [review]

File: .circleci/config.yml [hunk 22]
```diff
@@ -12,16 +12,10 @@ workflows:
             branches:
               ignore: releases
```
[the remark this answers is not in the document]
Comment (mfranz1) [reply .circleci/config.yml:14]: I'll think about it

File: .husky/pre-commit [hunk 23]
```diff
@@ -1,4 +1,4 @@
 #!/bin/sh
 . "$(dirname "$0")/_/husky.sh"
 
-npm run lint && npm run test:unit
+npm run lint
```
[the remark this answers is not in the document]
Comment (mfranz1) [reply .husky/pre-commit:4]: Good looks

File: src/app/admin/core/authorship.service.ts
[the remark this answers is not in the document]
Comment (mfranz1) [reply src/app/admin/core/authorship.service.ts:1]: The way we purpose `../core/..` is to support any Angular service or guard, in addition to a few client supporting components. The unique part of and `*.service.ts` file in a core is not always to interact with our API, it's a client service that controls a particular component, for an example. That is the case here in `admin/core` where most of these service files are controlling logic for client interactions. 

To fully answer your question, we should refactor to having one core within `src/app` however, I argue that moving every existing core to one core is a refactor out of scope of route updates for clark-service

File: src/app/core/auth-module/auth.service.ts
[the remark this answers is not in the document]
Comment (mfranz1) [reply src/app/core/auth-module/auth.service.ts:1]: In the login/register components in `src/app/auth`. The client needs a refactor

File: src/app/core/auth-module/auth.service.ts [hunk 24]
```diff
@@ -279,13 +295,10 @@ export class AuthService {
   async validateAndRefreshToken(): Promise<void> {
```
[the remark this answers is not in the document]
Comment (mfranz1) [reply src/app/core/auth-module/auth.service.ts:295]: Yee, the client needs to be refactored.

File: src/app/core/library-module/library.service.ts (hunk 6 above)
Comment (mfranz1) [reply src/app/core/library-module/library.service.ts:119]: Your guess is as good as mine. Client needs to be refactored

File: src/app/core/utility-module/utility.service.ts (hunk 20 above)
Comment (mfranz1) [reply src/app/core/utility-module/utility.service.ts:24]: No, there's not UI. We hope to add this to the client in the future.

File: src/app/core/access-group-module/access-group.service.ts [hunk 25]
```diff
@@ -0,0 +1,68 @@
...
+
+  /**
+   * Remove a user's privilege from a collection
+   *
+   * @param {string} collection the collection to remove
+   * @param {string} memberId the user who's privilege shall be revoked
+   * @returns {Promise<void>}
+   * @memberof UserService
+   */
+    async removeMember(collectionName: string, memberId: string): Promise<void> {
```
Comment (Cwagne17) [inline src/app/core/access-group-module/access-group.service.ts:44]: This function isn't being referenced anywhere

File: src/app/core/auth-module/auth.service.ts [hunk 26]
```diff
@@ -410,15 +386,13 @@ export class AuthService {
       const data = await this.encryptionService.encryptRSA(user);
       const response = await this.http
         .post<AuthUser & { tokens: Tokens }>(
-          environment.apiURL + '/users',
+          AUTH_ROUTES.REGISTER(),
           data,
```
[the remark this answers is not in the document]
Comment (Cwagne17) [reply src/app/core/auth-module/auth.service.ts:390]: It is the user on line 384 the register(user: any)

File: package.json [hunk 27]
```diff
@@ -7,8 +7,9 @@
     "ng": "ng",
     "start": "ng serve",
     "build": "ng build",
-    "test:unit": "jest --maxWorkers=2 --verbose --coverage uri-retriever featuredObjects",
     "lint": "npx ng lint clark",
+    "lint:fix": "npx ng lint clark --fix",
+    "prettier": "prettier --check .",
     "e2e": "ng e2e",
```
[the remark this answers is not in the document]
Comment (shanioffe) [reply package.json:13]: +1

**Review by shanioffe — APPROVED** [review]

File: src/app/admin/components/user-card/user-card.component.ts [hunk 28]
```diff
@@ -89,7 +89,7 @@ export class AdminUserCardComponent {
   }
 
   async canAddEvaluator() {
-    const roles = await this.privilegeService.getCollectionRoles(this.user.id);
+    const roles = await this.accessGroupsService.getUserAccessGroups(this.user.username);
```
Comment (pzalep1) [inline src/app/admin/components/user-card/user-card.component.ts:92]: Not using id?

File: src/app/admin/components/user-search-wrapper/user-search-wrapper.component.ts [hunk 29]
```diff
@@ -73,7 +73,6 @@ export class UserSearchWrapperComponent implements OnInit, OnDestroy {
         this.loading = false;
       }).catch(error => {
         this.toaster.error('Error!', 'There was an error fetching users. Please try again later.');
-        console.error(error);
```
Comment (pzalep1) [inline src/app/admin/components/user-search-wrapper/user-search-wrapper.component.ts:76]: Thank you for cleaning these up.

File: src/app/admin/pages/users/users.component.ts [hunk 30]
```diff
@@ -107,14 +115,21 @@ export class UsersComponent implements AfterViewInit {
...
+        username,
+        AccessGroups.REVIEWER,
+        this.activeCollection.abvName
+      )
+      .then(() => {
+        this.users = this.users.filter(x => x.username !== username);
+      })
+      .catch(error => {
+        this.toaster.error('Error!', 'Could not remove reviewer. Please try again later');
+        console.log(error);
```
Comment (pzalep1) [inline src/app/admin/pages/users/users.component.ts:131]: 🔥

File: src/app/auth/register/register.component.ts [hunk 31]
```diff
@@ -59,9 +59,9 @@ const EMAIL_REGEX =
     ])
   ]
 })
-export class RegisterComponent implements OnInit, OnDestroy{
+export class RegisterComponent implements OnInit, OnDestroy {
   private ngUnsubscribe = new Subject<void>();
-  ssoRedirect = environment.apiURL + '/google';
+  ssoRedirect = AUTH_ROUTES.GOOGLE_SIGNUP();
```
Comment (pzalep1) [inline src/app/auth/register/register.component.ts:64]: Nice!

File: src/app/core/access-group-module/access-group.service.ts [hunk 32]
```diff
@@ -0,0 +1,127 @@
...
+
+  private handleError(error: HttpErrorResponse) {
+    if (error.error instanceof ErrorEvent) {
+      // Client-side or network returned error
+      return throwError(error.error.message);
+    } else {
+      // API returned error
+      return throwError(error);
+    }
+  }
```
Comment (pzalep1) [inline src/app/core/access-group-module/access-group.service.ts:126]: I remember talking about an error service. Does that come later?

File: src/app/core/auth-module/auth.service.ts (hunk 26 above)
[the remark this answers is not in the document]
Comment (pzalep1) [reply src/app/core/auth-module/auth.service.ts:390]: ^

File: src/app/core/collection-module/collections.routes.ts [hunk 33]
```diff
@@ -0,0 +1,69 @@
...
+    GET_COLLECTION(abvName: string) {
+        return `${environment.apiURL}/collections/${encodeURIComponent(abvName)}`;
+    },
+    /**
+     * Request to retrieve the metrics of a collection
+     * @param abvName the abbreviated name of the collection
+     * @method GET
+     * @returns the metrics of the collection
+     */
+    GET_COLLECTION_METRICS(abvName: string) {
```
[the remark this answers is not in the document]
Comment (pzalep1) [reply src/app/core/collection-module/collections.routes.ts:35]: ^

File: src/app/core/collection-module/collections.service.ts [hunk 34]
```diff
@@ -42,12 +48,13 @@ export class CollectionService {
...
-              c.hasLogo = true;
-            });
-          } catch (_) {
+            await this.http.head('../../assets/images/collections/' + c.abvName + '.png')
+              .toPromise()
+              .then(() => {
+                c.hasLogo = true;
+              });
+          } catch (error) {
+            console.log(error);
```
Comment (pzalep1) [inline src/app/core/collection-module/collections.service.ts:57]: 🔥

File: src/app/core/feature-module/featured.service.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (pzalep1) [reply src/app/core/feature-module/featured.service.ts:136]: It is used on the featured admin page. It gets the learning objects that aren't featured.

File: src/app/core/interceptor/httpconfig.interceptor.ts [hunk 35]
```diff
@@ -20,9 +22,10 @@ export class HttpConfigInterceptor implements HttpInterceptor {
     request: HttpRequest<any>,
     next: HttpHandler
   ): Observable<HttpEvent<any>> {
+    console.log('intercepted request ... ');
```
Comment (pzalep1) [inline src/app/core/interceptor/httpconfig.interceptor.ts:25]: 🔥

File: src/app/core/library-module/library.service.ts (hunk 6 above)
Comment (pzalep1) [reply src/app/core/library-module/library.service.ts:119]: History: Library used to have a download button and we used to bundle everything on the fly.

File: src/app/core/metric-module/metric.routes.ts (hunk 8 above)
Comment (pzalep1) [reply src/app/core/metric-module/metric.routes.ts:29]: This is probably related to the weird 2 stats routes we had in library and learning object service.

**Review by pzalep1 — COMMENTED** [review]
Looks good! Just a few comments and questions.
